File: go_algorithms/algorithms/RowByColumnParallel.go
package algorithms

import (
	"fmt"
	"sync"
	"time"
)

func RowByColumnParallel(A, B [][]int) ([][]int, int64) {
	fmt.Println("Método: RowByColumnParallel")

	startTime := time.Now().UnixNano()

	size := len(A)
	C := make([][]int, size)
	for i := range C {
		C[i] = make([]int, size)
	}

	var wg sync.WaitGroup
	wg.Add(size * size)

	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			go func(row, col int) {
				defer wg.Done()
				sum := 0
				for k := 0; k < size; k++ {
					sum += A[row][k] * B[k][col]
				}
				C[row][col] = sum
			}(i, j)
		}
	}

	wg.Wait()

	endTime := time.Now().UnixNano()
	executionTime := endTime - startTime
	return C, executionTime
}

File: python_algorithms/algorithms/RowByColumnParallel.py
import numpy as np
import threading
import time

def Multiply(A, B):
    print("Método: RowByColumnParallel")

    start_time = time.time_ns()

    size = len(A)
    C = np.zeros((size, size), dtype=np.int64)

    def calculate_element(row, col):
        sum_val = 0
        for k in range(size):
            sum_val += A[row][k] * B[k][col]
        C[row][col] = sum_val

    threads = []
    for i in range(size):
        for j in range(size):
            thread = threading.Thread(target=calculate_element, args=(i, j))
            thread.start()
            threads.append(thread)

    for thread in threads:
        thread.join()

    end_time = time.time_ns()
    execution_time = end_time - start_time
    return C, execution_time
